packages/utils/src/time.ts

```diff
@@ -1,98 +1,154 @@
+  // transactions of long-lived web pages.
+  const timeOrigin = Date.now() - performance.now();
 
-const performanceFallback: CrossPlatformPerformance = {
-  now(): number {
-    let now = Date.now() - INITIAL_TIME;
-    if (now < prevNow) {
-      now = prevNow;
-    }
-    prevNow = now;
-    return now;
-  },
-  timeOrigin: INITIAL_TIME,
-};
+  return {
+    now: () => performance.now(),
+    timeOrigin,
+  };
+}
 
-const crossPlatformPerformance: CrossPlatformPerformance = ((): CrossPlatformPerformance => {
-  // React Native's performance.now() starts with a gigantic offset, so we need to wrap it.
-  if (isReactNative()) {
-    return getReactNativePerformanceWrapper();
+/**
+ * Returns the native Performance API implementation from Node.js. Returns undefined in old Node.js versions that don't
+ * implement the API.
+ */
+function getNodePerformance(): Performance | undefined {
+  try {
+    const perfHooks = dynamicRequire(module, 'perf_hooks') as { performance: Performance };
+    return perfHooks.performance;
+  } catch (_) {
+    return undefined;
   }
+}
 
+/**
+ * The Performance API implementation for the current platform, if available.
+ */
+const platformPerformance = ((): Performance | undefined => {
   if (isNodeEnv()) {
-    try {
-      const perfHooks = dynamicRequire(module, 'perf_hooks') as { performance: CrossPlatformPerformance };
-      return perfHooks.performance;
-    } catch (_) {
-      return performanceFallback;
-    }
+    return getNodePerformance();
   }
+  return getBrowserPerformance();
+})();
 
-  const { performance } = getGlobalObject<Window>();
-
-  if (!performance || !performance.now) {
-    return performanceFallback;
-  }
+let timestampSource = dateTimestampSource;
 
-  // Polyfill for performance.timeOrigin.
-  //
-  // While performance.timing.navigationStart is deprecated in favor of performance.timeOrigin, performance.timeOrigin
-  // is not as widely supported. Namely, performance.timeOrigin is undefined in Safari as of writing.
-  if (performance.timeOrigin === undefined) {
-    // As of writing, performance.timing is not available in Web Workers in mainstream browsers, so it is not always a
-    // valid fallback. In the absence of a initial time provided by the browser, fallback to INITIAL_TIME.
-    // @ts-ignore ignored because timeOrigin is a readonly property but we want to override
-    // eslint-disable-next-line deprecation/deprecation
-    performance.timeOrigin = (performance.timing && performance.timing.navigationStart) || INITIAL_TIME;
-  }
+if (platformPerformance !== undefined) {
+  timestampSource = {
+    nowSeconds: () => (platformPerformance.timeOrigin + platformPerformance.now()) / 1000,
+  };
+}
 
-  return performance;
-})();
+/**
+ * Returns a timestamp in seconds since the UNIX epoch using the Date API.
+ */
+export const dateTimestampInSeconds = dateTimestampSource.nowSeconds.bind(dateTimestampSource);
```

Comment:
I don't see any gains from using `dateTimestampSource` instead of a simple named function. Unless we already know that we're going to add more methods there?

Reply:
Functions would of course work. I started this with documentation, and a common interface seemed like a reasonable way to explain that there are going to be multiple ways to generate a timestamp.
`Date.now` and `performance.now` are also methods, not global functions.

When exporting from `utils` though, exporting a function keeps with the existing usage (`timestampWithMs`) in call sites.